text/0136-mobile-tracing-without-performance-v-2.md

```diff
@@ -0,0 +1,76 @@
+SDKs prefer the `traceID` of transactions bound to the scope over the `PropagationContext`. So
+when performance is disabled, or no transaction is bound to the scope, mobile SDKs use the same
+`traceId` and `spanId` for all captured events. This can lead to traces with hundreds of
+unrelated events confusing users. JS addressed this recently by updating the `PropagationContext`
+based on routes, see [Ensure browser traceId lifetime works as expected](https://github.com/getsentry/sentry-javascript/issues/11599).
+
+# Options Considered
+
+## Option 1: Update `PropagationContext` based on screens/routes <a name="option-1"></a>
+
+Mobile SDKs base the lifetime of the `traceId` of the `PropagationContext` on screens/routes,
+which is similar to a route on JavaScript. Mobile SDKs already report the screen name automatically
+via `view_names` with the [app context](https://develop.sentry.dev/sdk/event-payloads/contexts/#app-context)
+and use the same information for the name of screen load transactions, which the screen load
+starfish module uses. Whenever the screen name changes automatically or with a yet to be defined
+[manual API](https://www.notion.so/sentry/Specs-Screens-API-084d773272f24f57aeb622c07619264e),
+mobile SDKs must renew the `traceId` of the `PropagationContext`. The screen load transaction
+and subsequent events on the same screen must use the same `traceId`. When the app moves to the
+foreground after being in the background for longer than 30 seconds, which is the same approach
+mobile SDKs use for determining the end of a session, mobile SDKs renew the `traceId` of the
+PropagationContext. If the app stays in the background for shorter or equal to 30 seconds,
+mobile SDKs must not renew the `traceId` of the PropagationContext when the app moves again to
+the foreground. When a span starts, SDKs should use the traceID on the PropagationContext, but
+SDKs make an exception to this rule to map all spans related to a screen/route to the same traceID.
+For example, when a span leads to a new screen, SDKs should use the traceID of the new screen.
+
+### Pros <a name="option-1-pros"></a>
+
+1. Similar to [JavaScript]((https://github.com/getsentry/sentry-javascript/issues/11599)) updating
+it based on routes, so it should be easy to implement for React-Native.
+2. Works for spans first, as all spans get added to one trace per screen.
+
+### Cons <a name="option-1-cons"></a>
+
+1. For single-screen applications such as social networks, the lifetime of a trace could still be
+long, and multiple unrelated events could be mapped to one trace.
+2. For applications running for a long time in the background, such as running apps, the lifetime of
+a trace could still be long, and multiple unrelated events could be mapped to one trace.
+3. It doesn’t work well for declarative UI frameworks as Jetpack Compose and SwiftUI for which the
+SDKs can’t reliably automatically detect when apps load a new screen.
+
+# Drawbacks
```

Comment:
@Lms24, please clarify why this wasn't a problem before. I don't understand how this proposed change here will cause this.

Reply:
I might be missing a bit of context around how Tracing without Performance and the PropagationContext is implemented in mobile SDKs. 

If the proposed change is purely intended for TwP scenarios and does not affect the overall trace lifetime in case a root span/transaction _is_ started ("tracing with performance") I think we're good. That is because for TwP, we defer the sampling decision to the downstream service (i.e. send `sentry-trace` headers without a sampled flag).

In JS however, we changed the trace lifetime not just for TwP but in general, leading to scenarios like the one above. To illustrate further, why this is problematic, I'm gonna adjust the example a bit from above

1. Initial Pageload transaction is sampled by rolling the dice
2. PropagationContext stores positive sampling decision
3. application, still on the same page, makes an http request to a downstream service and propagates the `sentry-trace` header _with_ the positive sampling decision, forcing the downstream service to positively sample their transaction.
4. repeat 3 a lot of times (e.g. an application auto-refreshing some state every 5s) and you have a lot of sampled transactions because one initial transaction was sampled positively in the FE.

So even without an active transaction, we'd still propagate a forced sampled flag to downstream services.

Does this make sense?

Reply:
Thanks for the detailed explanation. Yes, that makes sense, but I guess in the long run, you should have a roughly equal amount of transactions. It shouldn't matter if you roll the dice once for 10 transactions or every time for each transaction. If you roll the dice often enough, an equal amount of transactions should be captured.

Reply:
Not necessarily unfortunately. This would only hold up if the sample rates on client and server were the same. If users have lower sample rates on the server, they would send significantly less server-side transactions with the previous implementation.

I tried verifying this with a small script: https://gist.github.com/Lms24/9a631295aef58cf22fb8f5307953335c

Reply:
> When a span starts, SDKs should use the traceID on the PropagationContext

We should say explicitly to use only the `traceId`, but not the sampling decision of the PropagationContext.
Regardless of client/server side sampling, i think we would break the sampling in general, as it doesn't apply to the PropagationContext. The `sampler` function is particularly problematic imho